store/sqlstore/channel_store.go

```diff
@@ -3132,3 +3401,287 @@ func (s SqlChannelStore) GroupSyncedChannelCount() (int64, *model.AppError) {
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			// clean previous preference favorites (to support bi-directional sync)
+			sql, args, _ = s.getQueryBuilder().Delete("Preferences").Where(sq.And{
+				sq.Eq{"ChannelId": oldCategory.Channels},
+				sq.Eq{"Preferences.Category": model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL},
+			}).ToSql()
+
+			if _, err = transaction.Exec(sql, args...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+		var channels []interface{}
+		var preferences []interface{}
+		runningOrder := 0
+		for _, channelID := range oldCategory.Channels {
+			channels = append(channels, &model.SidebarChannel{
+				ChannelId:  channelID,
+				CategoryId: category.Id,
+				SortOrder:  int64(runningOrder),
+				UserId:     userId,
+			})
+			preferences = append(preferences, &model.Preference{
+				Name:     channelID,
+				UserId:   userId,
+				Category: model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL,
+				Value:    "true",
+			})
+			runningOrder += model.MinimalSidebarSortDistance
+		}
+
+		if err = transaction.Insert(channels...); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			if err = transaction.Insert(preferences...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	}
+
+	if err = transaction.Commit(); err != nil {
+		return nil, model.NewAppError("SqlChannelStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.commit_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return categories, nil
+}
+
+// UpdateSidebarChannelByPreference is called when the Preference table is being updated to keep SidebarCategories in sync
+func (s SqlChannelStore) UpdateSidebarChannelByPreference(preference *model.Preference) *model.AppError {
+	if preference.Category != model.PREFERENCE_CATEGORY_DIRECT_CHANNEL_SHOW && preference.Category != model.PREFERENCE_CATEGORY_GROUP_CHANNEL_SHOW && preference.Category != model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL {
+		return nil
+	}
+	params := map[string]interface{}{
+		"UserId":       preference.UserId,
+		"ChannelId":    preference.Name,
+		"CategoryType": "",
+	}
+
+	switch preference.Category {
+	case model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL:
+		params["CategoryType"] = model.SidebarCategoryFavorites
+	case model.PREFERENCE_CATEGORY_DIRECT_CHANNEL_SHOW:
+	case model.PREFERENCE_CATEGORY_GROUP_CHANNEL_SHOW:
+		params["CategoryType"] = model.SidebarCategoryDirectMessages
```

Comment:
Was this here before? I don't know if having the DM/GM preferences update the SidebarChannel entries for the DMs category works since there's unfortunately ways for a DM/GM to become visible without updating this preference.

I think we might want to remove this part for now and just focus on keeping the favourites in sync between preferences and the sidebar for now since there's still more that goes into the visibility of DMs and GMs.

Reply:
@hmhealey  I am not sure I understand. Why should we remove this? What ways are there for DMs to become visible without this part?
If we remove this - how do you propose we keep them in sync?

Reply:
DMs can unfortunately become visible on the client when you receive a message in one, but the Preference isn't sent back to the server until the user actually views the channel. This is currently all controlled on the client-side I forgot about that when writing the spec for this, and I meant to mention it to you, but I forgot.

I think doing it the way you're doing it makes sense in the long run, but we might need to compensate for this somewhere (either on the client by adding to the received categories in this case or on the server by updating the preferences when a message is received in a DM), so I'll file a followup to just ensure that that behaviour works before this is shipped.

Reply:
ok, discussed this with @hmhealey offline, for now this function will only handle favorites